critiquebrainz/db/review.py

```diff
@@ -0,0 +1,608 @@
+
+    query = sqlalchemy.text("""
+        SELECT COUNT(*)
+          FROM review
+            {}
+        """.format(filterstr))
+
+    with db.engine.connect() as connection:
+        result = connection.execute(query, filter_data)
+        count = result.fetchone()[0]
+    sort = kwargs.pop("sort", None)
+    order_by_clause = str()
+
+    if sort == 'rating':
+        order_by_clause = """
+            ORDER BY rating DESC
+        """
+    elif sort == 'created':
+        order_by_clause = """
+            ORDER BY created DESC
+        """
+    elif sort == 'random':
+        order_by_clause = """
+            ORDER BY RANDOM()
+        """
+    # Note that all revisions' votes are considered in these ratings
+    query = sqlalchemy.text("""
+        SELECT review.id,
+               entity_id,
+               entity_type,
+               edits,
+               is_draft,
+               is_hidden,
+               license_id,
+               language,
+               review.user_id,
+               display_name,
+               show_gravatar,
+               is_blocked,
+               email,
+               "user".created as user_created,
+               musicbrainz_id,
+               MIN(revision.timestamp) as created,
+               SUM(
+                CASE
+                 WHEN vote='t' THEN 1
+                 ELSE 0
+                END
+               ) AS votes_positive_count,
+               SUM(
+                CASE
+                 WHEN vote='f' THEN 1
+                 ELSE 0
+                END
+               ) AS votes_negative_count,
+               SUM(
+                CASE
+                 WHEN vote = 't' THEN 1
+                 WHEN vote = 'f' THEN -1
+                 WHEN vote IS NULL THEN 0
+                END
+               ) AS rating,
+               latest_revision.id as latest_revision_id,
+               latest_revision.timestamp as latest_revision_timestamp,
+               latest_revision.text as text,
+               source,
+               source_url,
+               license.full_name,
+               license.info_url
+          FROM review
+          JOIN revision
+            ON review.id = revision.review_id
+     LEFT JOIN vote
+            ON vote.revision_id = revision.id
+          JOIN "user"
+            ON review.user_id = "user".id
+          JOIN license
+            ON license.id = license_id
+          JOIN (
+                revision
+                JOIN (
+                  SELECT review.id AS review_uuid,
+                         MAX(timestamp) AS latest_timestamp
+                    FROM review
+                    JOIN revision
+                      ON review.id = review_id
+                GROUP BY review.id
+                     ) AS latest
+                  ON latest.review_uuid = revision.review_id
+                 AND latest.latest_timestamp = revision.timestamp
+               ) AS latest_revision
+            ON review.id = latest_revision.review_id
+            {}
```

Comment:
Placeholder names should be explicitly specified: `{where_clause}` instead of `{}`, etc.